relay/pkg/block/forwarder.go

```diff
@@ -0,0 +1,110 @@
+package block
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/go-log"
+	"github.com/keep-network/tbtc/relay/pkg/btc"
+	"github.com/keep-network/tbtc/relay/pkg/chain"
+)
+
+const (
+	// Size of the headers queue.
+	headersQueueSize = 50
+
+	// Maximum size of processed headers batch.
+	headersBatchSize = 5
+
+	// Maximum time for which the pulling process will wait for a single header
+	// to be delivered by the headers queue.
+	headerTimeout = 1 * time.Second
+
+	// Block duration of a Bitcoin difficulty epoch.
+	difficultyEpochDuration = 2016
+
+	// Duration for which the forwarder should rest after performing
+	// a push action.
+	forwarderPushingSleepTime = 45 * time.Second
+)
+
+var logger = log.Logger("relay-block-forwarder")
+
+// Forwarder takes blocks from the Bitcoin chain and forwards them to the
+// given host chain.
+type Forwarder struct {
```

Comment:
Thoughts on renaming it to `Relay`?

Reply:
I'm not sure as the entire subdirectory is named `Relay`, we have a `Relay` contract and we have `chain.Relay` 😁  I think `Forwader` is a better name for things we are doing here. I always considered `Relay` to be a high-level concept which consists of BTC and ETH connectivity, block forwarder, and auxiliary stuff.

Reply:
I think this is fine, it is the same situation as with `keep-core/pkg/beacon` or `keep-ecdsa/pkg/ecdsa` where the "heart" package of the client is named after the protocol. There is much more logic happening inside this package than just forwarding: `updateBestHeader`, `findLastCommonAncestor` - this code is about how the relay works.

Reply:
Technically, `updateBestHeader` and `findLastCommonAncestor` are parts of the forwarding logic as well. Also, If we rename `block.Forwarder` to `relay.Relay` then according to your comment, the future auction stuff should be also placed in `relay` package as it's how the relay will work. This seems wrong to me. There is also the thing that `relay.Relay` will use `chain.Relay` which seems weird. Finally, I think `keep-core` and `keep-ecdsa` cases are not the same. The `beacon` package is something different than the whole repo name. Same in `keep-ecdsa` where the core logic is placed in `ecdsa/tss`.

Reply:
> Technically, updateBestHeader and findLastCommonAncestor are parts of the forwarding logic as well.

Right, and that's one of the things that are IMO confusing because it is not just forwarding from one place to another (just like HTTP forward for instance).

> Also, If we rename block.Forwarder to relay.Relay then according to your comment, the future auction stuff should be also placed in relay package as it's how the relay will work

No, the actions should be placed in `node.go` (or in `auctions` package referenced from `node.go`) and depending on the result of the auction, the client will start the relay or not.

> There is also the thing that relay.Relay will use chain.Relay which seems weird.

IMO it's consistent - we have a relay off-chain logic that is using relay on-chain implementation. `chain.Relay` is just an on-chain part of the interface.

> Finally, I think keep-core and keep-ecdsa cases are not the same. The beacon package is something different than the whole repo name. Same in keep-ecdsa where the core logic is placed in ecdsa/tss.

For the beacon, it's a kind of a historical decision that "core" is beacon + grant and staking contracts. If we had a separate repository for grant and staking contracts, the repository with the client would be named `keep-beacon` (this may still happen one day). For `keep-ecdsa`, we have a separate `tss` directory because we use third party implementation, but if we were using our own, it would live in `ecdsa`.

My order of preferences is:
- `pkg/relay` with `relay.go` and `Start()`
- `pkg/headers` with `headers.go` and  `StartRelay()`or `Transmit()`
- `pkg/headers` with `headers.go` and `RunForwarder()`

I don't think `pkg/block` is a good choice given that we work on headers and I don't think `forwarder.go` is a good choice because we do much more than just forwarding from one place to another.

Reply:
Yeah, I agree with the remark about `pkg/block`, this was a starting point which now can be narrowed to `pkg/header`. In that case, we can make `header.Relay` placed in `pkg/header/relay.go` exposing a `StartRelay` function. This is a small mix of your propositions but I think it's the best reflection of what we are doing here. Of course, I propose to do this as a follow-up PR after the entire stack lands on master. I'll leave a TODO. WDYT?

Reply:
Yeah, that's the best outcome we could get here IMO when I look at it again 👌 . Let's leave a TODO.

Reply:
Here it is: https://github.com/keep-network/tbtc/commit/bce322f8b5881c410789c9c5f7a885815dbcc5f2